src/module/system/damage/weapon.ts

```diff
@@ -391,14 +391,38 @@ class WeaponDamagePF2e {
             }
         }
 
-        // Roll notes
+        // Critical specialization effects
+        const critSpecEffect = ((): CritSpecEffect => {
+            if (!weapon.isOfType("weapon")) return [];
+
+            // If an alternate critical specialization effect is available, apply it only if there is also a
+            // qualifying non-alternate
+            const critSpecs = actor.synthetics.criticalSpecalizations;
+            const standard = critSpecs.standard
+                .map((cs) => cs(weapon, options))
+                .filter((ce): ce is CritSpecEffect => !!ce)
+                .pop();
+            const alternate = critSpecs.alternate
+                .map((cs) => cs(weapon, options))
+                .filter((ce): ce is CritSpecEffect => !!ce)
+                .pop();
```

Comment:
Oh I see, you wanted the last one _that didn't evaluate to null_.  It's still possible to not process all the ones before the final one.

```
const standard = critSpecs.standard.reduceRight((res, cs) => res ?? cs?.(weapon, options), null);
```
This starts from the end and as soon as a cs that non-false is found it stops evaluating them.

Reply:
If you see opportunities to micro-optimize, so make a PR